### DISCUSSION TITLE
Judgment in taiko is bit different between stable and lazer

### DISCUSSION THREAD

**Discussion by cdwcgt** [ppy/osu#30414] [Q&A] [answered]
osr file:[Hidden is fun - Nog - Aether Ritual [Inner Oni] (2024-10-24) Taiko.zip](https://github.com/user-attachments/files/17509631/Hidden.is.fun.-.Nog.-.Aether.Ritual.Inner.Oni.2024-10-24.Taiko.zip)
played(recored) in stable

In Taiko, if you press two buttons at the same time, [only one object will be judged](https://github.com/ppy/osu/blob/85d292ae5741e9460b7c7f1b231dc8e8b5818256/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableHit.cs#L120-L121).
but in stable is bit different, Sometimes double pressing works, but sometimes it doesn't

I did some research on the osr file attached above, I found that some time points were recorded with two frames.
![one time two frame](https://github.com/user-attachments/assets/7cbc7aef-938c-4b6a-9299-2567619ff0af)
as picture, there are frames where only one key is pressed and both keys are pressed at the same time.
double pressing this point will cause a miss on lazer, but not in stable even at gameplay

and for another 
![one frame only](https://github.com/user-attachments/assets/b6a5e516-14a1-4f66-94b6-0025116cb61a)
At this point, there is only one frame that presses two keys.
and double pressing will cause miss both stable and lazer

[cdwcgt playing Abydos High School Countermeasure Council - Kagayaki Summer Days (Cut Ver.) (komasy) [Pleasant Summer] (2024-10-24_21-43) (2).osr.zip](https://github.com/user-attachments/files/17509709/cdwcgt.playing.Abydos.High.School.Countermeasure.Council.-.Kagayaki.Summer.Days.Cut.Ver.komasy.Pleasant.Summer.2024-10-24_21-43.2.osr.zip)
played(recored) in lazer

![one time two frame](https://github.com/user-attachments/assets/b232b997-52a8-4c3c-93ae-78a454ecd0e7)
this cause miss in lazer both at replay and gameplay but stable didn't at replay

**Comment by bdach** [comment]
From the description it seems that stable's behaviour is weird and arbitrary and we don't want it? What would you even be expecting to happen here?

**Reply by cdwcgt** [reply]
This means that taiko players will feel the difference between lazer and stable. The intuitive experience is that stable can do what lazer can't (It will not miss in stable but will definitely miss in lazer) and don't know why

I just put this phenomenon here for people who need it, such as potential lazer taiko players.

I don't know how to handle this situation, so I set the discussion type to QA

**Comment by peppy** [comment]
> In Taiko, if you press two buttons at the same time, [only one object will be judged](https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableHit.cs#L118).

Are you sure you're understanding this correctly? That code is not stopping another object from being judged as it's *local to one object already*. If it were to do that, a slow update frame where a left and right button press arrived together would result in missing a hit that should have been hit.

IMO such a case should result in two objects both being hit if the key presses were valid.

**Reply by cdwcgt** [reply]
because `lastPressHandleTime == Time.Current`, so previous object will block the input
In the test it is the expected behavior
https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Taiko.Tests/Judgements/TestSceneHitJudgements.cs#L58

**Reply by peppy** [reply]
Hmm, that's pretty suspicious. I'm pretty sure that's not how TnT handles this scenario...

**Reply by peppy** [reply, marked as answer]
Tracking as an issue: https://github.com/ppy/osu/issues/30621
